File: project4/src/go/ising.go
package ising

import (
	"fmt"
	"math"
	"math/rand"
)

type Lattice struct {
	Spin [][]int8		// spin values (may be +1 or -1)
	L int 					// dimension of lattice
	Energy float64
	Tot_magnetization int64
	Accepted_configurations int64
}

func wraparound(i, offset, len int) int {
  return (len + (i + offset)) % len
}

func find_energy(lat_ptr *Lattice, J float64) {
	L := lat_ptr.L
	A := lat_ptr.Spin

	//var E float64 = 0
	var E int64 = 0


	for i := 0; i < L; i++ {
		for j := 0; j < L; j++ {
			E += int64(A[i][j] * A[(i+1) % L][j])    // horizontal neighbour
			E += int64(A[i][j] * A[i][(j+1) % L])    // vertical neighbour
		}
	}
	lat_ptr.Energy = -J*float64(E)
}

func find_tot_magnetization(lat_ptr *Lattice) {
	L := lat_ptr.L
	A := lat_ptr.Spin


	var tot_magnetization int64
	for i := 0; i < L; i++ {
			for j := 0; j < L; j++ {
					tot_magnetization += int64(A[i][j])
			}
	}
	lat_ptr.Tot_magnetization = tot_magnetization}

/* function to compute the change of energy, given that lat.data[i][j] is flipped */
func relative_change_of_energy(lat_ptr *Lattice, i int, j int) int {
	L := lat_ptr.L
	A := lat_ptr.Spin

	var E_old_up, E_old_down, E_old_left, E_old_right, E_old int8

	E_old_down =    A[i][j] * A[(i+1)%L][j]
	E_old_up =      A[i][j] * A[(L+i-1)%L][j]
	E_old_right =   A[i][j] * A[i][(j+1)%L]
	E_old_left =    A[i][j] * A[i][(L+j-1)%L]
	E_old = E_old_right + E_old_left + E_old_up + E_old_down

	/*
	the difference in energy is equal to
			dE = -J * (E_new - E_old)

	Since the change consists of flipping the sign of A[i][j],
			E_new = -E_old

	Implying
			dE = -J * (- 2 * E_old) = 2*J*E_old

	Since, only the relative value (scaled by a positive factor) is of
	importance, it is sufficient to return E_old
	*/
	return int(E_old)
}

func metropolis(lat_ptr *Lattice, sweeps int64, J float64, energies []float64,
	tot_magnetization []int64, r *rand.Rand, dE_cache []float64,
	save_every_nth int64) {
	L := lat_ptr.L
	lat := *lat_ptr
	spin := lat_ptr.Spin

	var ran float64
	var relative_dE int
	var accepted_configurations int64
	var pos_1d int64
	var sweep int64
	var L_sq int64 = int64(L*L)
	for sweep = 1; sweep <= sweeps; sweep++ {
		for count := 0; count < L*L; count++ {
			pos_1d = r.Int63n(L_sq)
			i := int(pos_1d / int64(L))
			j := int(pos_1d % int64(L))

			relative_dE = relative_change_of_energy(lat_ptr, i, j)
			dE := dE_cache[4+relative_dE]
			ran = r.Float64()

			if dE > ran {
				/* ACCEPT */
				spin[i][j] *= -1
				lat.Energy += 2*J*float64(relative_dE)
				lat.Tot_magnetization += int64(2*spin[i][j])
				accepted_configurations++
			}
		}
		if (sweep % save_every_nth) == 0 {
			energies[sweep/save_every_nth] = lat.Energy
			tot_magnetization[sweep/save_every_nth] = lat.Tot_magnetization
		}
	}
	lat.Accepted_configurations = accepted_configurations
	*lat_ptr = lat
}

func Solve(lat_ptr *Lattice, sweeps int64, J float64, T float64,
	energies []float64, tot_magnetization []int64, save_every_nth int64,
 	seed int64) {
	var beta float64 = 1 / T

	find_energy(lat_ptr, J)
	find_tot_magnetization(lat_ptr)

	dE_cache := make([]float64, 9)
	for i := 0; i <= 8; i += 2 {
		dE_cache[i] = math.Exp(-beta*float64(2*(i-4)))
	}

	// create new Rand
	var r = rand.New(rand.NewSource(seed))

	metropolis(lat_ptr, sweeps, J, energies, tot_magnetization, r, dE_cache,
				save_every_nth)
}

func Init_lattice(lat_ptr *Lattice, L int) {
}

func main() {
  fmt.Println("Test")
}

File: project4/src/julia/ising.jl
function wraparound(i, offset, len)
  return ((len + (i-1 + offset)) % len)+1
end

mutable struct Lattice
    Spin::Array{Int8, 2}    # spin values (may be +1 or -1)
    L::Int                  # dimension of lattice
    Energy::Float64
    Tot_magnetization::Int64
    Accepted_configurations::Int64
end

function find_energy(lat::Lattice, J)
    L = lat.L
    A = lat.Spin

    E = 0
    for i = 1:L
        for j = 1:L
            E += A[i,j] * A[wraparound(i, 1, L), j] # horizontal neighbour
            E += A[i,j] * A[i, wraparound(j, 1, L)] # vertical neighbour
        end
    end
    lat.Energy = -J*E
end

function find_tot_magnetization(lat::Lattice)
    L = lat.L
    A = lat.Spin

    tot_magnetization = 0
    for i = 1:L
        for j = 1:L
            tot_magnetization += A[i,j]
        end
    end

    lat.Tot_magnetization = tot_magnetization
end

function relative_change_of_energy(lat::Lattice, i::Int, j::Int)
    L = lat.L
    A = lat.Spin

    E_old_down =    A[i,j] * A[wraparound(i,1,L),  j]
    E_old_up =      A[i,j] * A[wraparound(i,-1,L), j]
    E_old_right =   A[i,j] * A[i, wraparound(j,1,L) ]
    E_old_left =    A[i,j] * A[i, wraparound(j,-1,L)]
    E_old = E_old_right + E_old_left + E_old_up + E_old_down

    #=
    the difference in energy is equal to
            dE = -J * (E_new - E_old)
    Since the change consists of flipping the sign of A[i][j],
            E_new = -E_old
    Implying
            dE = -J * (- 2 * E_old) = 2*J*E_old
    Since, only the relative value (scaled by a positive factor) is of
    importance, it is sufficient to return E_old
    =#
    return E_old
end

function metropolis(lat::Lattice, sweeps::Int64, J::Float64, energies::Array{Float64},
    tot_magnetization::Array{Int64}, r::MersenneTwister, dE_cache::Array{Float64},
    save_every_nth::Int64)
    L = lat.L
    spin = lat.Spin
    accepted_configurations = 0

    for sweep = 1:sweeps
        for count = 0:L*L
            pos_1d = rand(r, 0:L*L-1)
            i = div(pos_1d, L) + 1
            j = (pos_1d % L) + 1

            relative_dE = relative_change_of_energy(lat, i, j)
            dE = dE_cache[5+relative_dE]
            ran = rand(r, Float64)

            if dE > ran
                # ACCEPT
                spin[i,j] *= -1
                lat.Energy += 2*J*relative_dE
                lat.Tot_magnetization += 2*spin[i,j]
                accepted_configurations += 1
            end

        end

        if (sweep % save_every_nth) == 0
            energies[div(sweep, save_every_nth)] = lat.Energy
            tot_magnetization[div(sweep, save_every_nth)] = lat.Tot_magnetization
        end
    end

    lat.Accepted_configurations = accepted_configurations
end

function solve(lat::Lattice, sweeps::Int64, J::Float64, T::Float64,
    energies::Array{Float64}, tot_magnetization::Array{Int64}, save_every_nth::Int64,
    seed)
    beta = 1 / T

    find_energy(lat, J)
    find_tot_magnetization(lat)

    dE_cache = Array{Float64}(9)
    for i = 1:2:9
        dE_cache[i] = exp(-beta*2*(i-4))
    end

    # create new RNG
    r = MersenneTwister(seed)

    metropolis(lat, sweeps, J, energies, tot_magnetization, r, dE_cache,
                save_every_nth)
end
